File: rust/src/crates/journal-core/src/seal.rs
//! Forward Secure Sealing support for journal writers.
//!
//! Implements file-format sealing with deterministic synthetic keys,
//! matching systemd v260.1 HMAC byte ranges and tag object layout.
//!
//! HMAC byte ranges follow systemd journal-authenticate.c exactly:
//! - Header: signature..state, file_id..tail_entry_boot_id, seqnum_id..arena_size,
//!   data_hash_table_offset..tail_object_offset
//! - DATA: object header + hash + stored payload
//! - FIELD: object header + hash + payload
//! - ENTRY: object header + seqnum..end
//! - HASH_TABLE / ENTRY_ARRAY: object header only
//! - TAG: object header + seqnum + epoch (not the tag value itself)

use crate::error::{JournalError, Result};
use crate::file::ObjectType;
use crate::fss::{
    RECOMMENDED_SECPAR, RECOMMENDED_SEEDLEN, evolve, gen_mk, gen_state0, get_epoch, get_key,
};
use hmac::{Hmac, Mac};
use sha2::Sha256;

pub const TAG_LENGTH: usize = 256 / 8;

/// Type alias for HMAC-SHA256.
type HmacSha256 = Hmac<Sha256>;

/// Configures Forward Secure Sealing for a journal writer.
#[derive(Debug, Clone)]
pub struct SealOptions {
    pub seed: [u8; RECOMMENDED_SEEDLEN],
    pub interval_usec: u64,
    pub start_usec: u64,
}

impl SealOptions {
    pub fn new(seed: [u8; RECOMMENDED_SEEDLEN], interval_usec: u64, start_usec: u64) -> Self {
        Self {
            seed,
            interval_usec,
            start_usec,
        }
    }
}

/// Per-writer FSS+HMAC state.
pub struct SealState {
    fsprg_state: Vec<u8>,
    #[allow(dead_code)]
    msk: Vec<u8>,
    #[allow(dead_code)]
    seed: [u8; RECOMMENDED_SEEDLEN],
    interval: u64,
    start: u64,
    hmac: Option<HmacSha256>,
    hmac_running: bool,
}

impl SealState {
    pub fn new(opts: &SealOptions) -> Result<Self> {
        let (msk, mpk) = gen_mk(&opts.seed, RECOMMENDED_SECPAR);
        let state0 = gen_state0(&mpk, &opts.seed);
        Ok(Self {
            fsprg_state: state0,
            msk,
            seed: opts.seed,
            interval: opts.interval_usec,
            start: opts.start_usec,
            hmac: None,
            hmac_running: false,
        })
    }

    pub fn epoch(&self) -> u64 {
        get_epoch(&self.fsprg_state)
    }

    pub fn goal_epoch(&self, realtime: u64) -> Result<u64> {
        if self.start == 0 || self.interval == 0 {
            return Err(JournalError::FssVerificationError);
        }
        if realtime < self.start {
            return Err(JournalError::FssVerificationError);
        }
        Ok((realtime - self.start) / self.interval)
    }

    pub fn need_evolve(&self, realtime: u64) -> Result<bool> {
        let goal = self.goal_epoch(realtime)?;
        let epoch = self.epoch();
        if epoch > goal {
            return Err(JournalError::FssVerificationError);
        }
        Ok(epoch != goal)
    }

    pub fn hmac_start(&mut self) {
        if self.hmac_running {
            return;
        }
        let key = get_key(&self.fsprg_state, TAG_LENGTH, 0);
        self.hmac = Some(HmacSha256::new_from_slice(&key).expect("HMAC accepts any key length"));
        self.hmac_running = true;
    }

    pub fn hmac_write(&mut self, data: &[u8]) {
        self.hmac_start();
        if let Some(ref mut hmac) = self.hmac {
            hmac.update(data);
        }
    }

    pub fn hmac_reset(&mut self) {
        self.hmac_running = false;
        self.hmac = None;
    }

    pub fn hmac_finalize(&mut self) -> [u8; TAG_LENGTH] {
        let result = self
            .hmac
            .take()
            .expect("hmac_finalize called without active HMAC")
            .finalize()
            .into_bytes();
        self.hmac_running = false;
        let mut out = [0u8; TAG_LENGTH];
        out.copy_from_slice(&result);
        out
    }

    /// HMAC the immutable header byte ranges (systemd journal-authenticate.c:329-354).
    ///
    /// Four ranges from the serialized on-disk header:
    /// - signature through just before state (bytes 0..16)
    /// - file_id through just before tail_entry_boot_id (bytes 24..56)
    /// - seqnum_id through just before arena_size (bytes 72..96)
    /// - data_hash_table_offset through just before tail_object_offset (bytes 104..136)
    pub fn hmac_put_header_ranges(&mut self, header_bytes: &[u8]) {
        self.hmac_start();
        // signature + compatible_flags + incompatible_flags; state is excluded.
        self.hmac_write(&header_bytes[0..16]);

        // file_id + machine_id; tail_entry_boot_id is excluded.
        self.hmac_write(&header_bytes[24..56]);

        // seqnum_id + header_size; arena_size is excluded.
        self.hmac_write(&header_bytes[72..96]);

        // data/field hash-table offsets and sizes; tail_object_offset is excluded.
        self.hmac_write(&header_bytes[104..136]);
    }

    /// HMAC an object's immutable bytes based on its type.
    ///
    /// `object_bytes` must contain the full serialized on-disk object.
    /// `object_size` is the actual object size from the object header (not the buffer length).
    /// `is_compact` must be true for journals with HEADER_INCOMPATIBLE_COMPACT.
    ///
    /// Byte ranges follow systemd journal-authenticate.c:267-327:
    /// - All objects: object header (up to payload)
    /// - DATA: + hash + stored payload (offset 64 regular, 72 compact)
    /// - FIELD: + hash + payload
    /// - ENTRY: + seqnum through end
    /// - HASH_TABLE / ENTRY_ARRAY: nothing beyond header
    /// - TAG: + seqnum + epoch (not the tag value)
    pub fn hmac_put_object_bytes(
        &mut self,
        object_bytes: &[u8],
        typ: ObjectType,
        object_size: u64,
        is_compact: bool,
    ) {
        self.hmac_start();

        let object_header_size: u64 = 16; // type(1) + flags(1) + reserved(6) + size(8)

        // Object header is always HMAC'd for all types
        self.hmac_write(&object_bytes[..object_header_size as usize]);

        match typ {
            ObjectType::Data => {
                // systemd journal-authenticate.c:293-294:
                //   hmac(data.hash) + hmac(payload from payload_offset to end)
                // Regular payload offset = offsetof(Object, data.regular.payload) = 64
                // Compact payload offset = offsetof(Object, data.compact.payload) = 72
                let hash_offset = object_header_size as usize;
                self.hmac_write(&object_bytes[hash_offset..hash_offset + 8]);

                let payload_offset: u64 = if is_compact { 72 } else { 64 };
                if object_size > payload_offset {
                    let payload_size = (object_size - payload_offset) as usize;
                    let start = payload_offset as usize;
                    if start + payload_size <= object_bytes.len() {
                        self.hmac_write(&object_bytes[start..start + payload_size]);
                    }
                }
            }
            ObjectType::Field => {
                // hash (8 bytes at offset 16) + payload
                let hash_offset = object_header_size as usize;
                self.hmac_write(&object_bytes[hash_offset..hash_offset + 8]);

                // FieldObjectHeader: object_header(16) + hash(8) + next_hash_offset(8) +
                //   head_data_offset(8) = 40
                // Payload starts at offset 40
                let payload_offset: u64 = 40;
                if object_size > payload_offset {
                    let payload_size = (object_size - payload_offset) as usize;
                    let start = payload_offset as usize;
                    if start + payload_size <= object_bytes.len() {
                        self.hmac_write(&object_bytes[start..start + payload_size]);
                    }
                }
            }
            ObjectType::Entry => {
                // Everything from seqnum onward
                // EntryObjectHeader: object_header(16) + seqnum(8) + xor_hash(8) +
                //   boot_id(16) + monotonic(8) + realtime(8) + n_items(8) = 72
                // But systemd HMACs from &o->entry.seqnum, which is at offset 16 (after object header)
                // Length: object_size - object_header_size
                if object_size > object_header_size {
                    let rest_size = (object_size - object_header_size) as usize;
                    let start = object_header_size as usize;
                    if start + rest_size <= object_bytes.len() {
                        self.hmac_write(&object_bytes[start..start + rest_size]);
                    }
                }
            }
            ObjectType::DataHashTable | ObjectType::FieldHashTable | ObjectType::EntryArray => {
                // Nothing beyond object header: everything is mutable
            }
            ObjectType::Tag => {
                // seqnum (8 bytes) + epoch (8 bytes), not the tag value itself
                // TagObjectHeader: object_header(16) + seqnum(8) + epoch(8) + tag(32) = 64
                // seqnum is at offset 16, epoch at offset 24
                let seqnum_offset = object_header_size as usize;
                self.hmac_write(&object_bytes[seqnum_offset..seqnum_offset + 8]);
                self.hmac_write(&object_bytes[seqnum_offset + 8..seqnum_offset + 16]);
            }
            ObjectType::Unused => {}
        }
    }

    pub fn evolve_state(&mut self) {
        self.fsprg_state = evolve(&self.fsprg_state);
    }
}

File: go/journal/seal.go
package journal

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"hash"
)

const (
	tagLength                  = 256 / 8
	objectTypeTag              = 7
	compatibleSealed           = 1 << 0
	compatibleSealedContinuous = 1 << 2
)

// SealOptions configures Forward Secure Sealing for a journal writer.
// Seed must be exactly 12 bytes.
type SealOptions struct {
	Seed         []byte
	IntervalUsec uint64
	// StartUsec is normalized to systemd's verification-key boundary:
	// floor(StartUsec / IntervalUsec) * IntervalUsec.
	StartUsec uint64
}

// sealState holds per-writer FSS+HMAC state.
type sealState struct {
	fsprgState  []byte
	msk         []byte
	seed        []byte
	interval    uint64
	start       uint64
	hmac        hash.Hash
	hmacRunning bool
}

func newSealState(opts SealOptions) (*sealState, error) {
	if len(opts.Seed) != fsprgRecommendedSeedlen {
		return nil, fmt.Errorf("seal seed must be %d bytes", fsprgRecommendedSeedlen)
	}
	if opts.IntervalUsec == 0 || opts.StartUsec < opts.IntervalUsec {
		return nil, fmt.Errorf("FSS start and interval must be set")
	}
	start := (opts.StartUsec / opts.IntervalUsec) * opts.IntervalUsec
	msk, mpk, err := fsprgGenMK(opts.Seed, fsprgRecommendedSecpar)
	if err != nil {
		return nil, err
	}
	state0 := fsprgGenState0(mpk, opts.Seed)
	return &sealState{
		fsprgState: state0,
		msk:        msk,
		seed:       append([]byte(nil), opts.Seed...),
		interval:   opts.IntervalUsec,
		start:      start,
	}, nil
}

func (s *sealState) getEpoch() uint64 {
	return fsprgGetEpoch(s.fsprgState)
}

func (s *sealState) getGoalEpoch(realtime uint64) (uint64, error) {
	if s.start == 0 || s.interval == 0 {
		return 0, fmt.Errorf("FSS start or interval not set")
	}
	if realtime < s.start {
		return 0, fmt.Errorf("realtime before FSS start")
	}
	return (realtime - s.start) / s.interval, nil
}

func (s *sealState) needEvolve(realtime uint64) (bool, error) {
	goal, err := s.getGoalEpoch(realtime)
	if err != nil {
		return false, err
	}
	epoch := s.getEpoch()
	if epoch > goal {
		return false, fmt.Errorf("FSS epoch %d > goal %d", epoch, goal)
	}
	return epoch != goal, nil
}

func (s *sealState) hmacStart() {
	if s.hmacRunning {
		return
	}
	key := fsprgGetKey(s.fsprgState, tagLength, 0)
	s.hmac = hmac.New(sha256.New, key)
	s.hmacRunning = true
}

func (s *sealState) hmacWrite(p []byte) {
	s.hmacStart()
	s.hmac.Write(p)
}

func (s *sealState) hmacReset() {
	s.hmacRunning = false
	s.hmac = nil
}

func (s *sealState) hmacSum() []byte {
	return s.hmac.Sum(nil)
}

// tagObjectSize returns the on-disk size of a TAG object.
func tagObjectSize() uint64 {
	return uint64(objectHeaderSize + 8 + 8 + tagLength)
}

// appendTag creates a TAG object at offset, HMACs its header/seqnum/epoch,
// stores the current HMAC digest as the tag value, and resets the HMAC cycle.
// The caller must have ensured the object fits.
func (w *Writer) appendTag() error {
	if w.seal == nil {
		return nil
	}
	w.seal.hmacStart()

	offset := w.appendOffset
	size := tagObjectSize()
	seqnum := w.header.nTags + 1
	epoch := w.seal.getEpoch()

	buf := make([]byte, align8(size))
	putObjectHeader(buf[:objectHeaderSize], objectHeader{typ: objectTypeTag, size: size})
	binary.LittleEndian.PutUint64(buf[objectHeaderSize:objectHeaderSize+8], seqnum)
	binary.LittleEndian.PutUint64(buf[objectHeaderSize+8:objectHeaderSize+16], epoch)

	// HMAC object header + seqnum + epoch (exclude tag value)
	w.seal.hmacWrite(buf[:objectHeaderSize+16])

	// Store digest
	copy(buf[objectHeaderSize+16:objectHeaderSize+16+tagLength], w.seal.hmacSum())

	if err := w.writeObject(offset, buf); err != nil {
		return err
	}
	w.objectAdded(offset, size)
	w.header.nTags = seqnum
	w.seal.hmacReset()
	return nil
}

// appendFirstTag is called after hash table setup to create the initial TAG.
func (w *Writer) appendFirstTag() error {
	if w.seal == nil {
		return nil
	}
	if err := w.hmacPutHeader(); err != nil {
		return err
	}
	if err := w.hmacPutHashTableObject(w.header.fieldHashTableOffset - uint64(objectHeaderSize)); err != nil {
		return err
	}
	if err := w.hmacPutHashTableObject(w.header.dataHashTableOffset - uint64(objectHeaderSize)); err != nil {
		return err
	}
	return w.appendTag()
}

// maybeAppendTag checks whether the entry realtime crosses a sealing interval.
// If so, it appends a tag (finalizing the old epoch) and evolves the FSPRG.
func (w *Writer) maybeAppendTag(realtime uint64) error {
	if w.seal == nil {
		return nil
	}
	need, err := w.seal.needEvolve(realtime)
	if err != nil {
		return err
	}
	if !need {
		return nil
	}
	if err := w.appendTag(); err != nil {
		return err
	}
	// Evolve across intervals, appending intermediate tags.
	for {
		goal, err := w.seal.getGoalEpoch(realtime)
		if err != nil {
			return err
		}
		epoch := w.seal.getEpoch()
		if epoch >= goal {
			break
		}
		w.seal.fsprgState = fsprgEvolve(w.seal.fsprgState)
		if w.seal.getEpoch() < goal {
			if err := w.appendTag(); err != nil {
				return err
			}
		}
	}
	return nil
}

// hmacPutHeader updates the HMAC with the immutable header byte ranges.
func (w *Writer) hmacPutHeader() error {
	if w.seal == nil {
		return nil
	}
	w.seal.hmacStart()

	// signature through just before state: bytes 0-15
	w.seal.hmacWrite(w.headerBytes(0, 16))

	// file_id through just before tail_entry_boot_id: bytes 24-55
	w.seal.hmacWrite(w.headerBytes(24, 32))

	// seqnum_id through just before arena_size: bytes 72-95 (seqnum_id[16] + header_size[8])
	w.seal.hmacWrite(w.headerBytes(72, 24))

	// data_hash_table_offset through just before tail_object_offset: bytes 104-135
	w.seal.hmacWrite(w.headerBytes(104, 32))

	return nil
}

// hmacPutHashTableObject updates the HMAC with a hash table object read from the file.
func (w *Writer) hmacPutHashTableObject(objectStart uint64) error {
	if w.seal == nil {
		return nil
	}
	w.seal.hmacStart()
	buf := make([]byte, objectHeaderSize)
	if err := w.readAt(buf, objectStart); err != nil {
		return err
	}
	// Hash table objects: only object header is immutable
	w.seal.hmacWrite(buf)
	return nil
}

// hmacPutObject updates the HMAC with the immutable bytes of an object.
// The object type determines which bytes are HMAC'd.
func (w *Writer) hmacPutObject(objectStart uint64, typ uint8) error {
	if w.seal == nil {
		return nil
	}
	w.seal.hmacStart()

	// Object header (up to payload) is always HMAC'd
	buf := make([]byte, objectHeaderSize)
	if err := w.readAt(buf, objectStart); err != nil {
		return err
	}
	w.seal.hmacWrite(buf)

	switch typ {
	case objectTypeData:
		// hash (8 bytes) + payload
		hashBuf := make([]byte, 8)
		if err := w.readAt(hashBuf, objectStart+16); err != nil {
			return err
		}
		w.seal.hmacWrite(hashBuf)
		payloadOffset := w.dataPayloadOffset()
		payloadSize := binary.LittleEndian.Uint64(buf[8:16]) - payloadOffset
		if payloadSize > 0 {
			payload := make([]byte, payloadSize)
			if err := w.readAt(payload, objectStart+payloadOffset); err != nil {
				return err
			}
			w.seal.hmacWrite(payload)
		}
	case objectTypeField:
		// hash (8 bytes) + payload
		hashBuf := make([]byte, 8)
		if err := w.readAt(hashBuf, objectStart+16); err != nil {
			return err
		}
		w.seal.hmacWrite(hashBuf)
		payloadSize := binary.LittleEndian.Uint64(buf[8:16]) - uint64(fieldObjectHeaderSize)
		if payloadSize > 0 {
			payload := make([]byte, payloadSize)
			if err := w.readAt(payload, objectStart+fieldObjectHeaderSize); err != nil {
				return err
			}
			w.seal.hmacWrite(payload)
		}
	case objectTypeEntry:
		// everything from seqnum onward
		entrySize := binary.LittleEndian.Uint64(buf[8:16])
		if entrySize > uint64(objectHeaderSize) {
			rest := make([]byte, entrySize-uint64(objectHeaderSize))
			if err := w.readAt(rest, objectStart+objectHeaderSize); err != nil {
				return err
			}
			w.seal.hmacWrite(rest)
		}
	case objectTypeDataHashTable, objectTypeFieldHashTable, objectTypeEntryArray:
		// nothing beyond object header
	case objectTypeTag:
		// seqnum + epoch
		tagMeta := make([]byte, 16)
		if err := w.readAt(tagMeta, objectStart+objectHeaderSize); err != nil {
			return err
		}
		w.seal.hmacWrite(tagMeta)
	}
	return nil
}

// headerBytes returns a slice of the in-memory header at the given offset/length.
func (w *Writer) headerBytes(off, length uint64) []byte {
	buf := make([]byte, headerSize)
	putHeader(buf, w.header)
	return buf[off : off+length]
}
